### PR TITLE
[Breaking flag] fix(Backups): new badger SuperFlag, NumGoroutines option solves OOM crashes

### PR DESCRIPTION
**Pull request by karlmcguire**
(This PR is for DGRAPH-2953, paired with dgraph-io/badger#1656 which adds the Badger options.)

On master, Alpha nodes would crash during a backup on a machine with low memory (<6gb in my tests). This was because the default number of goroutines was [hardcoded as 16](https://github.com/dgraph-io/dgraph/blob/master/worker/backup_processor.go#L42) which is used to create [local threads](https://github.com/dgraph-io/dgraph/blob/master/worker/backup_processor.go#L68) and used to [overwrite `Stream.numGo`](https://github.com/dgraph-io/dgraph/blob/master/worker/backup_processor.go#L131). 

The solution is to add a `--badger "goroutines=8; ..."` SuperFlag to Alpha and anything else that might want to set Badger options. So far I've also added it to `bulk/run.go`, but that might be too much for this PR. The command line flags for Badger should be the same across each tool, for consistency.

The question now is how we should calculate defaults. The default `goroutines` value of 8 seems to be working fine on a 4gb container, but should we lessen it for machines with less memory?

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7387)

### CODE REVIEW

File: dgraph/cmd/alpha/run.go [hunk 1]
```diff
@@ -108,11 +108,15 @@ they form a Raft group and provide synchronous replication.
...
-		"[none, zstd:level, snappy] Specifies the compression algorithm and the compression"+
-			"level (if applicable) for the postings directory. none would disable compression,"+
-			" while zstd:1 would set zstd compression at level 1.")
+	flag.String("badger", worker.BadgerDefaults,
+		`Various badger options.
+	goroutines=N provides the number of goroutines to use in badger.Stream.
+	compression=[none, zstd:level, snappy] specifies the compression algorithm and the compression 
+		level (if applicable) for the postings directory. "none" would disable compression, while 
+		"zstd:1" would set zstd compression at level 1.
+	cache_mb=N total size of cache (in MB) per shard in the reducer.
```
[the remark this answers is not in the document]
Comment (jarifibrahim) [reply dgraph/cmd/alpha/run.go:117]: We also have the posting list cache. We can move both the badger caches inside the `--badger` flag and keep the posting list size flag outside.

Something like `dgraph --badger="cache_mb=100; cache_percentage=50,50" --posting-cache-mb=50`

File: worker/backup_processor.go [hunk 2]
```diff
@@ -38,8 +38,7 @@ import (
 )
 
 const (
-	// backupNumGo is the number of go routines used by the backup stream writer.
-	backupNumGo = 16
+	BadgerDefaults = "goroutines=8; compression=snappy; cache_mb=64; cache_percentage=70,30;"
```
Comment (jarifibrahim) [inline worker/backup_processor.go:41]: You can keep this in the `worker/server_state.go` file.

Comment (jarifibrahim) [reply worker/backup_processor.go:41]: Keep the cache size to 1024. 64 mb might be too small.

File: dgraph/cmd/alpha/run.go [hunk 3]
```diff
@@ -108,11 +108,12 @@ they form a Raft group and provide synchronous replication.
...
 	flag.String("tmp", "t", "Directory to store temporary buffers.")
 
-	// Options around how to set up Badger.
-	flag.String("badger.compression", "snappy",
-		"[none, zstd:level, snappy] Specifies the compression algorithm and the compression"+
-			"level (if applicable) for the postings directory. none would disable compression,"+
-			" while zstd:1 would set zstd compression at level 1.")
+	flag.String("badger", worker.BadgerDefaults,
+		`Various badger options.
+	goroutines=N provides the number of goroutines to use in badger.Stream.
```
Comment (karlmcguire) [inline dgraph/cmd/alpha/run.go:113]: We need this `goroutines` flag in Alpha for [creating local threads during backups](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-e64b7a1608a78718294f5229733bcdfc7a383ecb5a359dd82ab35883874a94cbR63), and Badger uses the `goroutines` value as the default `Stream.numGo`.

File: dgraph/cmd/alpha/run.go [hunk 4]
```diff
@@ -108,11 +108,12 @@ they form a Raft group and provide synchronous replication.
...
+	compression=[none, zstd:level, snappy] specifies the compression algorithm and the compression
```
Comment (karlmcguire) [inline dgraph/cmd/alpha/run.go:114]: We need this `compression` flag in Alpha for [setting the compression type](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-a72445362e52218163e2d2f572b946ff3aea61f8675aa235a5dab3227ac6168eR610) of Badger.

File: dgraph/cmd/bulk/run.go [hunk 5]
```diff
@@ -115,21 +115,22 @@ func init() {
...
-		"[none, zstd:level, snappy] Specifies the compression algorithm and the compression"+
-			"level (if applicable) for the postings directory. none would disable compression,"+
-			" while zstd:1 would set zstd compression at level 1.")
-	flag.Int64("badger.cache_mb", 64, "Total size of cache (in MB) per shard in reducer.")
-	flag.String("badger.cache_percentage", "70,30",
-		"Cache percentages summing up to 100 for various caches"+
-			" (FORMAT: BlockCacheSize, IndexCacheSize).")
+	flag.String("badger", worker.BadgerDefaults,
+		`Various badger options.
+	compression=[none, zstd:level, snappy] specifies the compression algorithm and the compression
```
Comment (karlmcguire) [inline dgraph/cmd/bulk/run.go:120]: We need this `compression` flag in Bulk for [setting the compression type](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R133) of Badger (similar to how Alpha uses it).

File: dgraph/cmd/bulk/run.go [hunk 6]
```diff
@@ -115,21 +115,22 @@ func init() {
...
+		level (if applicable) for the postings directory. "none" would disable compression, while
+		"zstd:1" would set zstd compression at level 1.
+	cache_mb=N total size of cache (in MB) per shard in the reducer.
```
Comment (karlmcguire) [inline dgraph/cmd/bulk/run.go:123]: We need this `cache_mb` flag in Bulk for [setting the Badger cache size](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R170). The default value is set in [`worker/server_state.go`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-02ceb0db5e4dd0958bdba03c2fdee17011b75a2fc34cdaec7d232a828d7bfda8R34).

File: dgraph/cmd/bulk/run.go [hunk 7]
```diff
@@ -115,21 +115,22 @@ func init() {
...
+	cache_percentage=N cache percentages summing up to 100 for various caches.
```
Comment (karlmcguire) [inline dgraph/cmd/bulk/run.go:124]: We need this `cache_percentage` flag in Bulk for [setting the BlockCacheSize and IndexCacheSize](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R172).

File: worker/backup_processor.go [hunk 8]
```diff
@@ -65,7 +60,7 @@ func NewBackupProcessor(db *badger.DB, req *pb.BackupRequest) *BackupProcessor {
 	bp := &BackupProcessor{
 		DB:      db,
 		Request: req,
-		threads: make([]*threadLocal, backupNumGo),
+		threads: make([]*threadLocal, x.WorkerConfig.Badger.GetUint64("goroutines")),
```
Comment (karlmcguire) [inline worker/backup_processor.go:63]: TODO: Figure out how to calculate this at runtime so we never have OOM errors with default flags?

File: worker/server_state.go [hunk 9]
```diff
@@ -30,6 +30,10 @@ import (
 	"github.com/golang/glog"
 )
 
+const (
+	BadgerDefaults = "goroutines=8; compression=snappy; cache_mb=64; cache_percentage=70,30"
```
Comment (karlmcguire) [inline worker/server_state.go:34]: Some of these defaults aren't used by Alpha, which is kind of confusing...

We use [`goroutines`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-a72445362e52218163e2d2f572b946ff3aea61f8675aa235a5dab3227ac6168eR113) and [`compression`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-a72445362e52218163e2d2f572b946ff3aea61f8675aa235a5dab3227ac6168eR114) in Alpha.

We use [`compression`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R120), [`cache_mb`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R123), and [`cache_percentage`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R124) in Bulk.

**Review by manishrjain — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7387)* status: 0 of 5 files reviewed, 14 unresolved discussions (waiting on @karlmcguire and @vvbalaji-dgraph)

___
*[dgraph/cmd/alpha/run.go, line 609 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7387#-MSddWd60nJPxI374FZS:-MSddWd8BLu7Z4w7gzJP:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/1d65d59695b55405d4e53002a8f0a02a03a01134/dgraph/cmd/alpha/run.go#L609)):*
> ```Go
> 	walCache := (cachePercent[3] * (totalCache << 20)) / 100
> 
> 	badger := x.NewSuperFlag(Alpha.Conf.GetString("badger")).MergeAndCheckDefault(worker.BadgerDefaults)
> ```

100 chars.
___
*[dgraph/cmd/bulk/run.go, line 120 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7387#-MT6Bek37Boz2J1yy1qS:-MT6Bek37Boz2J1yy1qT:b-d3zjqa) ([raw file](https://github.com/dgraph-io/dgraph/blob/1d65d59695b55405d4e53002a8f0a02a03a01134/dgraph/cmd/bulk/run.go#L120)):*
> ```Go
> 	flag.String("badger", worker.BadgerDefaults,
> 		`Various badger options.
> 	compression=[none, zstd:level, snappy] specifies the compression algorithm and the compression
> ```

This string can be in one place. Also, we should be using Badger options directly. Whatever options Badger already has, should do directly to Badger.
___
*[dgraph/cmd/bulk/run.go, line 132 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7387#-MT6BrnJABme4z4SnDGR:-MT6BrnJABme4z4SnDGS:b-1ex4r) ([raw file](https://github.com/dgraph-io/dgraph/blob/1d65d59695b55405d4e53002a8f0a02a03a01134/dgraph/cmd/bulk/run.go#L132)):*
> ```Go
> 
> func run() {
> 	badger := x.NewSuperFlag(Bulk.Conf.GetString("badger")).MergeAndCheckDefault(worker.BadgerDefaults)
> ```

100 chars.
___
*[worker/server_state.go, line 34 at r2](https://reviewable.io/reviews/dgraph-io/dgraph/7387#-MSYJ34KNekuFqMrO4qh-r2-34:-MT6C-NL4Unah-ZCpyqA:b-jcolot) ([raw file](https://github.com/dgraph-io/dgraph/blob/1d65d59695b55405d4e53002a8f0a02a03a01134/worker/server_state.go#L34)):*
<details><summary><i>Previously, karlmcguire (Karl McGuire) wrote…</i></summary><blockquote>

Some of these defaults aren't used by Alpha, which is kind of confusing...

We use [`goroutines`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-a72445362e52218163e2d2f572b946ff3aea61f8675aa235a5dab3227ac6168eR113) and [`compression`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-a72445362e52218163e2d2f572b946ff3aea61f8675aa235a5dab3227ac6168eR114) in Alpha.

We use [`compression`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R120), [`cache_mb`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R123), and [`cache_percentage`](https://github.com/dgraph-io/dgraph/pull/7387/files#diff-76237e26c3354ceb21e16c8c34877ddb485e468d4eb711ba35477f35f9018e96R124) in Bulk. 

It doesn't make sense to split up the `BadgerDefaults` across packages, but it might be worth adding a comment explaining which packages use which options.
</blockquote></details>

Let's not change the default value of cache.

**Comment by karlmcguire** [comment]
> This string can be in one place.

Where? We have Badger flags for Alpha and Bulk, both using different flags. Alpha used to expose `compression` and `goroutine`, Bulk used to expose `compression`, `cache_mb`, and `cache_percentage`. If we get rid of `cache_mb` and `cache_percentage` then we could probably just use the same `compression` and `goroutine` flag string for Alpha and Bulk.

> Also, we should be using Badger options directly. Whatever options Badger already has, should do directly to Badger.

I'm not sure what this means.           

> Let's not change the default value of cache.

We're currently using `cache_mb` and `cache_percentage` in Bulk [here](https://github.com/dgraph-io/dgraph/blob/master/dgraph/cmd/bulk/run.go#L124-L127) for the [block and index caches](https://github.com/dgraph-io/dgraph/blob/master/dgraph/cmd/bulk/run.go#L173-L179). Should we not expose these flags?

**Pull request merged**